---  
    title: "Codex Example"  
    author: "Gene"  
    date: "`r format(Sys.Date())`"
    output:  
      html_document:  
        keep_md: true 
---  

#install and load packages
```{r message=FALSE}
#install.packages("devtools", repos = "http://archive.linux.duke.edu/cran/", quiet=T)
library(devtools, quietly=T)
#install_github("yuchaojiang/CODEX/package")
library(CODEX, quietly=T)
#install_github("WES.1KG.WUGSC")
#library(WES.1KG.WUGSC, quietly=T) # Load Toy data from the 1000 Genomes Project.
#source("https://bioconductor.org/biocLite.R")
#biocLite("WES.1KG.WUGSC")
library(WES.1KG.WUGSC, quietly=T)
```

#Get bam file directories, sample names from .txt file, and exonic positions from .bed file.

```{r}
dirPath <- system.file("extdata", package = "WES.1KG.WUGSC")
bamFile <- list.files(dirPath, pattern = '*.bam$')
bamdir <- file.path(dirPath, bamFile)
sampname <- as.matrix(read.table(file.path(dirPath, "sampname")))
bedFile <- file.path(dirPath, "chr22_400_to_500.bed")
chr <- 22
bambedObj <- getbambed(bamdir = bamdir, bedFile = bedFile, 
                       sampname = sampname, projectname = "CODEX_demo", chr)
bamdir <- bambedObj$bamdir; sampname <- bambedObj$sampname
ref <- bambedObj$ref; projectname <- bambedObj$projectname; chr <- bambedObj$chr

bambedObj$ref[1:10,]
```


#Get depth of coverage for each exon across all samples from whole exome sequencing files.
```{r message=FALSE}
coverageObj <- getcoverage(bambedObj, mapqthres = 20)
Y <- coverageObj$Y; readlength <- coverageObj$readlength
plot(x= seq_along(Y), y = log(Y), xlab="Exon", ylab= "Log Coverage", pch=20)
```



#Compute GC content for each exon. Will be later used in QC procedure and normalization.

```{r}
gc <- getgc(chr, ref)
plot(x= seq_along(gc), y = gc, xlab="Exon", ylab= "GC%", pch=20)
```


#Compute mappability for each exon. To save running time, take values from pre-computed results. Will be later used in QC procedure.

```{r}
mapp <- getmapp(chr, ref)
plot(x= seq_along(mapp), y =mapp, xlab="Exon", ylab="Mappability", pch=20)
```


#Apply a quality control procedure to the depth of coverage matrix both sample-wise and exon-wise before normalization.


```{r}
qcObj <- qc(Y, sampname, chr, ref, mapp, gc, cov_thresh = c(20, 4000), 
    length_thresh = c(20, 2000), mapp_thresh = 0.9, gc_thresh = c(20, 80))
Y_qc <- qcObj$Y_qc; sampname_qc <- qcObj$sampname_qc; gc_qc <- qcObj$gc_qc
mapp_qc <- qcObj$mapp_qc; ref_qc <- qcObj$ref_qc; qcmat <- qcObj$qcmat
write.table(qcmat, file = paste(projectname, '_', chr, '_qcmat', '.txt', sep=''), sep='\t', quote=FALSE, row.names=FALSE)
qcmat[1:10,]
```


#Fit a Poisson log-linear model that normalizes the read depth data from whole exome sequencing. Includes terms that specifically remove biases due to GC content, exon capture and amplification efficiency, and latent systemic artifacts.

```{r message=FALSE}
normObj <- normalize(Y_qc, gc_qc, K = 1:9)
Yhat <- normObj$Yhat; AIC <- normObj$AIC; BIC <- normObj$BIC
RSS <- normObj$RSS; K <- normObj$K
```


#Plot RSS, AIC, and BIC by number of latent variables.  

```{r}
# filename <- paste(projectname, "_", chr, "_choiceofK", ".pdf", sep = "")
# Kmax <- length(AIC)
par(mfrow = c(1, 3))
plot(K, RSS, type = "b", xlab = "Number of latent variables", pch=20)
plot(K, AIC, type = "b", xlab = "Number of latent variables", pch=20)
plot(K, BIC, type = "b", xlab = "Number of latent variables", pch=20)
```


#Recursive segmentation algorithm for CNV detection and genotyping, using normalized read depth from whole exome sequencing.

```{r message=FALSE}
optK = K[which.max(BIC)]
finalcall <- segment(Y_qc, Yhat, optK = optK, K = K, sampname_qc,
    ref_qc, chr, lmax = 200, mode = "integer")
```

```{r}
finalcall
```







